### DISCUSSION TITLE
openHASP on RPi zero

### DISCUSSION THREAD

**Discussion by DolceTriade** [HASwitchPlate/openHASP#324] [Q&A] [answered]
Hello,

I have an older RPi zero with a 3.5in touch screen lying around. Would it be possible to port the code to run on that?

**Comment by fvanroie** [comment, marked as answer]
It's possible, but it hasn't been ported to RPi. The code can be built to run on Linux as a desktop app through SDL2.
But for a Pi you'd probably want to use a framebuffer and evdev touch driver... neither have been ported and tested.
So, it depends on the time and skills.

If you have an SPI RPi display, it might be easier to get an ESP32-One from Waveshare since it has the same pin header.

**Reply by DolceTriade** [reply]
Hmm, yes. I tried to build the app for Linux, but it seemed like it was most useful for testing layouts since things like the http server were not working and if we wanted them to work, we'd need to add support for the Linux networking. I'd be willing to take a stab however, I just wanted to understand if there would need to be any major architectural changes or whether things could be done like adding a new driver...

**Reply by fvanroie** [reply]
Yes it's mostly used for testing. Only the paho MQTTclient is implemented for POSIX.

The other services aren't necessary _per se_, since you already have a desktop environment to edit, copy and configure files.
On Linux you could install nginx, ssh, samba, ... to achieve the same goal. Do you really need the HTTP server?

There is the route to go for Linux alternatives or implement support for [compiling Arduino sketches for RPi](https://www.deviceplus.com/raspberry-pi/how-to-run-arduino-sketches-on-raspberry-pi/)...

IMHO it only makes sense if the Pi would drive larger touchscreens and LVGL can run directly from the command-line instead of needing a full blown desktop environment. The code is quite modular so adding a framebuffer and evdev driver should be possible without too much effort.

Also making the paho client more robust, so it can handle network and MQTT reconnections gracefully would be a plus.

**Reply by nagyrobi** [reply]
Yeah I'd love to see a solution for HD/FullHD resolution...

**Reply by DolceTriade** [reply]
imo, if you have the capability to run HD, you are powerful enough to run a browser and can just run Home Assistant or something in Kiosk mode. The only reason this is appealing is because the RPi Zero isn't powerful enough to run a browser with Home Assistant properly. I'll take a stab at it!

**Reply by nagyrobi** [reply]
The problem with Home Assistant's main UI is that it's very very heavy and unstable on the long term. Too many unused features squashed in. You need something like a Core I3/I5 and a recent browser (kept up to date) to keep it alive, and restart it (programatically) every 2-3 days. An Android tabled can't be refurbished, any hardware older than 2-3 years will be completely hogged down, freezing and crashing etc. Been there, seen that.

OpenHASP can run stable and safe for months without requring a restart. I'd vote for a dedicated piece of hardware, completely off the internet, just communicating with HA, nothing else. Not requiring any fancy hardware acceleration for screens fading in and out etc.

And there are many other systems beside Home Assistant which could benefit from a dedicated, third party, MQTT-based bigger screen.

**Reply by DolceTriade** [reply]
I see...
